netdog: remove conditional compilation

jmt-lab opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:** 
Closes #3618

Removes conditional compilation of netdog for wicked and systemd networkd. This change moves our build system to build two variants of the netdog rpm: one for wicked and one for systemd-networkd. To facilitate this change the following was done:

* Migrate `#[cfg(net_backend = ]` to `#[cfg(feature = `
* Add two cargo features to sources/api/netdgo: 'wicked' and 'systemd-networkd'
* Extract netdog to its own package folder (packages/netdog)
* Extract netdog rpm building to netdog.spec
* Create two subpackages one netdog-wicked  and netdog-systemd-networkd  which are selected via image-features
* Migrated netdog to utilize the dogtag hostname detectors


**Testing done:**
Built and published aws-ecs-1 (wicked) and aws-ecs-2 (systemd-networkd) and verified network accessibility and use in a ecs cluster.


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

sam-berning commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,96 @@
+%cargo_prep
+
+%build
+mkdir bin
```
nit: is this directory used anywhere?

jmt-lab replied:
For the netdog binary itself i believe, needs to be created before hand

cbgbt commented on sources/api/netdog/Cargo.toml:
```diff
@@ -1,13 +1,18 @@
 [package]
 name = "netdog"
-version = "0.1.0"
+version = "0.1.1"
```
We've historically not done much management of subcomponent versioning. I'm supportive of using meaningful versioning for these changes, especially for subcomponents that will get their own RPM.

cbgbt commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+
+# RPM BuildRequires
+[build-dependencies]
+glibc = { path = "../glibc" }
```
We prefer to add newlines at EOF of text files per posix standards. This is configurable in most editors.

I'd suggest doing an interactive re-base to re-add missing newlines.

jmt-lab replied:
You got it

cbgbt commented on sources/api/netdog/Cargo.toml:
```diff
@@ -9,7 +9,7 @@ publish = false
 [features]
-default = ["systemd-networkd"]
+default = ["wicked"]
 wicked = []
```
Are the unit tests something we can fix here? I think advocating for `systemd-networkd` as the default makes sense.

jmt-lab replied:
will fix

sam-berning commented on sources/api/netdog/Cargo.toml:
```diff
@@ -1,13 +1,18 @@
 exclude = ["README.md"]
 
+[features]
+default = ["wicked"]
```
it feels wrong to me that wicked is the default feature flag, but we're using systemd-networkd as the default version of netdog otherwise.

I see the change was made to pass the unit tests, is it possible to rewrite the tests to either work with systemd-networkd or to specifically use the wicked features?

jmt-lab replied:
Good catch, I forgot to add the unit test change, I can modify the unit test to work with both, as systemd-networkd only adds extra lines (if i change the test to only verify that the first line is the expected it should pass on both and we can flip the default)

sam-berning commented on variants/aws-ecs-2/Cargo.toml:
```diff
@@ -42,6 +43,7 @@ path = "../variants.rs"
 # core
 release = { path = "../../packages/release" }
 kernel-6_1 = { path = "../../packages/kernel-6.1" }
+netdog = {path = "../../packages/netdog" }
```
why is netdog added as a build dependency only on this variant?

jmt-lab replied:
Somehow this escaped my remove all

sam-berning commented on variants/aws-ecs-1-nvidia/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ included-packages = [
 # core
     "release",
     "kernel-5.10",
+    "netdog-wicked",
```
checking my understanding, when a variant directly includes this subpackage, it overwrites the link created by `netdog-systemd-networkd`, right? are we sure that the packages are always installed in the correct order?

jmt-lab replied:
Yes essentially, so netdog defines two subpackages:

netdog-systemd-networkd which provides netdog:1 (epoch 1)
netdog-wicked which provides netdog:0 (epoch 0)

So way to think of the above without it the pseudo dnf install resolution would go dnf install netdog (find the default latest epoch) this would auto resolve to netdog-systemd-networkd since it provides netdog with highest epoch

However if you manually say install netdog-wicked it will ensure that netdog is provided by netdog:0 or netdog-wicked

sam-berning replied:
got it, thanks for the explanation

cbgbt commented on packages/netdog/Cargo.lock:
```diff
@@ -0,0 +1,14 @@
+# This file is automatically @generated by Cargo.
```
I don't believe that we tend to commit lockfiles for individual packages.

cbgbt commented on packages/settings-motd/Cargo.lock:
```diff
@@ -0,0 +1,14 @@
+# This file is automatically @generated by Cargo.
```
This similarly seems like it shouldn't be here.

cbgbt commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,96 @@
+Source7: generate-network-config.service
+
+BuildRequires: %{_cross_os}glibc-devel
+Requires: %{_cross_os}netdog
+%description
```
nit: mind adding whitespace between these?

jmt-lab replied:
Sure

cbgbt commented on packages/os/Cargo.lock:
```diff
@@ -0,0 +1,14 @@
+# This file is automatically @generated by Cargo.
```
Similar extraneous lockfile here.

cbgbt commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,96 @@
+install -p -m 0644 %{S:2} %{S:3} %{S:7} %{buildroot}%{_cross_unitdir}
+install -d %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+install -p -m 0644 %{S:6} %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+%if %{with vmware_platform}
```
I think we may want this file to be included in yet another package which we include in `vmware` variants. Right now this package still requires conditional compilation on vmware.

jmt-lab replied:
LIke a third subpackage netdog-vmware which we then require in the vmware variants ontop of the normal

cbgbt replied:
After looking at the specific service, while it is networking related, I wouldn't claim that it's necessarily part of netdog.

I think what we can do for now is move this back into `os` and opt to include it in a variant-specific `vmware` package

jmt-lab replied:
like adding a vmware-config subpackage in os?

yeazelm replied:
Although it bloats the number of rpms a tad, I think I like the idea of a `netdog-vmware-config` rpm and possibly moving the rest of the shared config into its own rpm and only basically swapping the resulting binary and `Requires` directives for the resulting rpms. That would allow others to "bring their own" netdog binary without having to remember to find and collect all the right configuration files needed. This is more preference and can be persuaded either way on the approach. I don't know if we really expect anyone to ever mess with and provide their own netdog but I think its more about setting a pattern for how we group config files that might be shared among mutually exclusive binaries. If I'm hijacking this thread we can move it away from here too.

cbgbt commented on variants/aws-dev/Cargo.toml:
```diff
@@ -28,6 +27,7 @@ included-packages = [
 # core
     "release",
     "kernel-6.1",
+    
```
nit: I'd suggest pruning (or standardizing) the added whitespace, since it's not consistent between files.

cbgbt commented on packages/netdog/netdog.spec:
```diff
@@ -90,7 +90,7 @@ install -p -m 0755 ${HOME}/.cache/wicked/%{__cargo_target}/release/netdog %{buil
 
 %post -n %{_cross_os}netdog-wicked -p <lua>
-fs.link("%{_cross_bindir}/netdog-wicked", "%{_cross_bindir}/netdog")
+posix.link("%{_cross_bindir}/netdog-wicked", "%{_cross_bindir}/netdog")
```
stylistic nit that might appear strangely in the GitHub UI. I'd suggest rebasing this change to the one that introduces the new RPM, since it's otherwise irrelevant here.

cbgbt commented on packages/release/Cargo.lock:
```diff
@@ -0,0 +1,570 @@
+# This file is automatically @generated by Cargo.
```
I wonder if these are being generated by an editor plugin like rust-analyzer? We may want to concoct a .gitignore for these entries.

jmt-lab replied:
Thats exactly what is happening

cbgbt reviewed (commented):
Nice work! Please pardon any midair collisions with @sam-berning. I think we reviewed concurrently.

yeazelm commented on sources/api/netdog/src/cli/mod.rs:
```diff
@@ -228,7 +228,9 @@ mod tests {
-        assert_eq!(std::fs::read_to_string(&fake_file).unwrap(), expected);
+        assert!(std::fs::read_to_string(&fake_file)
+            .unwrap()
+            .starts_with(expected));
```
This actually changes the semantics a bit no? We are checking for precisely the content listed above, did you need to make the `starts_with` change to get it to pass?

jmt-lab replied:
systemd-networkd generates the correct line however, it also generates extra lines for ipv6 settings. Considering this test was looking for the first line i assumed keeping it the same would be preferable

yeazelm commented on variants/aws-dev/Cargo.toml:
```diff
@@ -12,7 +12,6 @@ grub-set-private-var = true
 xfs-data-partition = true
 uefi-secure-boot = true
-systemd-networkd = true
 
```
So if I understand correctly, we are shifting the dependency to default into `systemd-networkd`? And we rely upon the netdog-wicked to bring in wicked.

jmt-lab replied:
correct

yeazelm commented on packages/.gitignore:
```diff
@@ -1,3 +1,4 @@
 *.patch.bz2
 *.src.rpm
 *.zip
+**/Cargo.lock
```
Was this intentional? I thought we actually did want to track `Cargo.lock` files?

jmt-lab replied:
This was in response to a preivous comment where my editor kept adding .lock files where not needed, sean suggested maybe we should setup a gitignore just for this level of Cargo.lock files

bcressey replied:
> This was in response to a preivous comment where my editor kept adding .lock files where not needed, sean suggested maybe we should setup a gitignore just for this level of Cargo.lock files

That sounds like weird editor behavior - is it something you can change locally? I haven't seen spurious `Cargo.lock` files show up in past pull requests.

webern commented on sources/api/netdog/Cargo.toml:
```diff
@@ -1,13 +1,18 @@
 exclude = ["README.md"]
 
+[features]
+default = ["systemd-networkd"]
+wicked = []
+systemd-networkd = []
```
It is not a great idea to have mutually exclusive features. It would be better to have a `wicked` feature, and to use `cfg(wicked)` and `cfg(not(wicked))`, otherwise `--all-features` will not compile.

But even so, this doesn't actually solve the problem of removing conditional compilation. We are still compiling conditionally. I would prefer it if the Rust package contained `bin/netdog` and `bin/netdog-wicked` which can both be compiled at the same time.

jmt-lab replied:
I went this approach to minimize the risk to the code and utilize standard rust/cargo approach for this. While it is true that generally you want to avoid mutually exclusive features, there is also not a valid case to compile with --all-features as a variant is not going to have BOTH wicked and systemd installed on it. Also the cfg(wicket) cfg(not(feature=wicked)) approach limits us to just these two and means if we add a third we'd have to readdress the feature flags entirely instead of adding a third feature. 

I do think I need to add a compile_error if a user tries to compile this with both feature flags, that is definitely a miss. I do understand where you are coming from on the not removing conditional compilation you are true the cargo crate still has conditional compilation. However, bottlerocket as a whole no longer does here as it compiles both netdog-systemd-networkd and netdog-wicked thus ensuring we build both versions.

jmt-lab commented:
Rolled back to handling the binary split in the rpm spec. It is more than sufficient and attempts to split the binary at cargo level fell upon issues with how netdog is structured that were not minimal to address

bcressey commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+publish = false
+build = "../build.rs"
+
+[lib]
+path = "../packages.rs"
```
You will need to use the [source groups](https://github.com/bottlerocket-os/twoliter/blob/40e21e03aecc18de092f9cd94bb045e6c3028d8d/tools/buildsys/src/manifest.rs#L12) feature to track changes under `sources`.

To avoid also triggering rebuilds of `os` when `netdog` changes, you'll need to relocate `sources/api/netdog` to `sources/netdog`.

jmt-lab replied:
👍

webern replied:
> To avoid also triggering rebuilds of os when netdog changes, you'll need to relocate sources/api/netdog to sources/netdog.

Why is this, do we have a watch on all of `sources/api`? Incidentally, I don't think I've ever seen a variant build in which os is not rebuilt.

jmt-lab replied:
This was done

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,85 @@
+%undefine _debugsource_packages
+
+Name: %{_cross_os}netdog
+Version: 0.1.1
```
I'd prefer to keep a `0.0` version here - I don't want people thinking they have to increment this version whenever they make a change to `netdog`.

bcressey commented on packages/os/Cargo.toml:
```diff
@@ -35,6 +34,7 @@ glibc = { path = "../glibc" }
 
 # RPM Requires
 [dependencies]
+netdog = { path = "../netdog" }
```
The wicked and systemd-networkd dependencies down below this line should be cleaned up.

What's actually left in `os.spec` that requires `netdog`? I.e. is this a true dependency? If not we shouldn't list it.

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,85 @@
+%setup -T -c
+%cargo_prep
+
+%build
+mkdir bin
+
+echo "** Build Netdog Binaries"
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+    -p netdog \
+    --target-dir=${HOME}/.cache/networkd
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+    -p netdog \
+    --features wicked \
+    --target-dir=${HOME}/.cache/wicked
```
I'd like to be explicit in the first branch here and not default to one feature or the other:
```suggestion
%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
    -p netdog \
    --features networkd \
    --target-dir=${HOME}/.cache/networkd
%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
    -p netdog \
    --features wicked \
    --target-dir=${HOME}/.cache/wicked
```

jmt-lab replied:
So you would like me to return the netdog package to having two features instead of one, Matt's main argument for having the one is that --all-features would still work (albeit will only test the wicked branch)

webern replied:
I have a preference for non-mutually-exclusive features (the way you have it now), but perhaps you can use `--features default` or `--default-features=true` to make this more self documenting.

```sugestion
# Build with systemd-networkd as the backend
%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
    -p netdog \
    --features default \
    --default-features=true \
    --target-dir=${HOME}/.cache/networkd

# Build with wicked as the backend
%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
    -p netdog \
    --features wicked \
    --default-features=false
    --target-dir=${HOME}/.cache/wicked
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,85 @@
+install -d %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+install -p -m 0644 %{S:6} %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+install -p -m 0755 ${HOME}/.cache/networkd/%{__cargo_target}/release/netdog %{buildroot}%{_cross_bindir}/netdog-systemd-networkd
+install -p -m 0755 ${HOME}/.cache/wicked/%{__cargo_target}/release/netdog %{buildroot}%{_cross_bindir}/netdog-wicked
```
Style-wise, I would prefer to create a directory and populate it before moving to the next:
```
install -d %{buildroot}%{_cross_bindir}
install -p -m 0755 ${HOME}/.cache/networkd/%{__cargo_target}/release/netdog %{buildroot}%{_cross_bindir}/netdog-systemd-networkd
install -p -m 0755 ${HOME}/.cache/networkd/%{__cargo_target}/release/netdog %{buildroot}%{_cross_bindir}/netdog-systemd-networkd

install -d %{buildroot}%{_cross_tmpfilesdir}
install -p -m 0644 %{S:4} %{buildroot}%{_cross_tmpfilesdir}/netdog.conf

...
```

Also it would be good to number the sources so that within a "type" (systemd unit, template) the numbers are consecutive. That pairs with the above advice because it makes the lines more obviously correct - if all units are in the range from `%{S:10}` to `{S:19}` then the reviewer doesn't have to do a lot of scrolling back and forth to figure out which source is being installed.

(`release.spec` is structured this way for this reason.)

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,85 @@
+
+%files -n %{_cross_os}netdog-systemd-networkd
+%{_cross_bindir}/netdog-systemd-networkd
+%{_cross_tmpfilesdir}/netdog.conf
```
Some of these files are needed for both `wicked` and `networkd` backends. Convention is to place these in a package like `%{_cross_os}netdog-common` that both packages require. Then you don't have "files listed twice" RPM warnings.

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,85 @@
+posix.link("%{_cross_bindir}/netdog-wicked", "%{_cross_bindir}/netdog")
+
+%post -n %{_cross_os}netdog-systemd-networkd -p <lua>
+posix.link("%{_cross_bindir}/netdog-systemd-networkd", "%{_cross_bindir}/netdog")
```
scriptlets come between `%install` and the first `%files` section

jmt-lab commented:
Reason for return of using rpmspec:

The reason for this was a bunch of roadblock updates i ran into with splitting the binaries out:

* I first attempted a pretty straight forward split where i copied [main.rs](http://main.rs/) to two bin/networkd.rs and bin/wicked.rs files, this worked fine however leads to very large duplicate files since netdog currently re-uses its [main.rs](http://main.rs/) to set a large amount of static consts used throughout the crate
* I then tried a hack by just symlinking [main.rs](http://main.rs/) twice to avoid cargo complaining about two targets using the same [main.rs](http://main.rs/) This did not propagate well into git and our ci cd
* I then tried to do a lib and bin split where i took all the static const and most of the shared code to be in [lib.rs](http://lib.rs/) and only moved the async main to the separated files, however this led to a collision with how netdog's multiple modules use private error messages with no shared top-type (meaning build error on visibility) this is fixable but would require alot ot making things public that i wasn't sure we wanted as such

As you can see above each approach i took to do the binary split at cargo time was bringing with it a bunch of more work when we had a functional solution. I decided to go back to a functional solution  and file a github issue (which i need to do) for us to refactor netdog in the future to handle both this and how the cfgs are too integrated into too many source files

bcressey commented on packages/os/os.spec:
```diff
@@ -48,20 +47,16 @@ Source114: bootstrap-containers@.service
 Source122: has-boot-ever-succeeded.service
-Source123: run-netdog.mount
-Source124: write-network-status.service
+Source123: disable-udp-offload.service
```
Let's move this to `netdog.spec` as well since it's networking related, or else to `release.spec`. `release.spec` is slightly more correct since that is where the `ethtool` requirement comes in.

bcressey commented on packages/os/os.spec:
```diff
@@ -99,6 +93,7 @@ Requires: %{_cross_os}xfscli
 Requires: %{_cross_os}thar-be-settings
 Requires: %{_cross_os}thar-be-updates
 Requires: %{_cross_os}updog
+Requires: %{_cross_os}netdog
```
This dependency should move to `release.spec`.

bcressey commented on packages/os/os.spec:
```diff
@@ -242,6 +226,11 @@ Summary: Bottlerocket certificates handler
+%package -n %{_cross_os}vmware-netconfig
+Summary: Bottlerocket vmware specific network configuration
+%description -n %{_cross_os}vmware-netconfig
+%{summary}.
```
This isn't necessary. `disable-udp-offload.service` already has `ConditionVirtualization=vmware` and won't run on other platforms, so we should unconditionally include it somewhere.

bcressey commented on packages/os/os.spec:
```diff
@@ -478,15 +466,10 @@ install -d %{buildroot}%{_cross_unitdir}
-install -d %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
-install -p -m 0644 %{S:12} %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
-%endif
+
```
nit: extra newline
```suggestion
```

bcressey commented on packages/os/os.spec:
```diff
@@ -511,17 +494,13 @@ install -p -m 0644 %{S:200} %{buildroot}%{_cross_tmpfilesdir}/migration.conf
 install -p -m 0644 %{S:202} %{buildroot}%{_cross_tmpfilesdir}/thar-be-updates.conf
 install -p -m 0644 %{S:203} %{buildroot}%{_cross_tmpfilesdir}/bootstrap-containers.conf
-install -p -m 0644 %{S:204} %{buildroot}%{_cross_tmpfilesdir}/netdog.conf
+
```
nit: remove added newline - we don't want extra whitespace between the different install blocks
```suggestion
```

bcressey commented on sources/api/netdog/Cargo.toml:
```diff
@@ -1,14 +1,19 @@
 exclude = ["README.md"]
 
+[features]
+default = []
+wicked = []
```
So that we can be explicit about the feature selection in `netdog.spec`, but still have CI test one of these paths:
```suggestion
default = ["networkd"]
networkd = []
wicked = []
```

But somewhere it should be an error to define both features.

bcressey commented on sources/api/netdog/src/cli/generate_net_config.rs:
```diff
@@ -7,7 +7,7 @@ use snafu::{OptionExt, ResultExt};
 use std::path::Path;
 
-#[cfg(net_backend = "systemd-networkd")]
+#[cfg(not(feature = "wicked"))]
```
```suggestion
#[cfg(feature = "networkd")]
```

bcressey commented on sources/api/netdog/src/main.rs:
```diff
@@ -24,7 +24,7 @@ valid example: `netdog.default-interface=eno1:dhcp4,dhcp6?`.
 supplementing any missing settings with DNS settings from the primary interface's DHCP lease.  It
 is meant to be used as a restart command for DNS API settings.
-*/
+ */
```
Is this happening from `cargo fmt`? If not I wouldn't include the change:
```suggestion
*/
```

bcressey commented on sources/api/netdog/src/networkd/config/netdev.rs:
```diff
@@ -142,7 +142,7 @@ impl Display for NetDevMacAddress {
 
 impl NetDevConfig {
-    const FILE_EXT: &str = "netdev";
+    const FILE_EXT: &'static str = "netdev";
```
Why did this change?

jmt-lab replied:
This is a requirement of upgrated rust sdk. I'm surprised we aren't being riddled with more of these errors, but when the rust update landed as part of 1.18 into develop and i upgraded the PR it started flagging this as an error (previously was a warning) All module global consts need to have a static lifetime if using a type that requires lifetimes such as &str. It is no longer implicit

bcressey commented on sources/api/netdog/src/cli/mod.rs:
```diff
@@ -366,17 +368,17 @@ mod error {
         },
 
-        #[cfg(net_backend = "systemd-networkd")]
+        #[cfg(not(feature = "wicked"))]
```
Switching from conditional compilation to a feature is a pretty big change - might be better to do it in its own commit.

bcressey commented on sources/api/netdog/src/cli/mod.rs:
```diff
@@ -228,7 +228,9 @@ mod tests {
         let expected = "-net.ipv4.conf.eno1.rp_filter = 2\n";
         write_interface_sysctl(interface, &fake_file).unwrap();
-        assert_eq!(std::fs::read_to_string(&fake_file).unwrap(), expected);
+        assert!(std::fs::read_to_string(&fake_file)
```
Was this an existing unit test failure that wasn't caught because we don't test the systemd-networkd path?

jmt-lab replied:
correct

bcressey commented on variants/aws-ecs-2/Cargo.toml:
```diff
@@ -12,7 +12,6 @@ grub-set-private-var = true
 unified-cgroup-hierarchy = true
 uefi-secure-boot = true
 xfs-data-partition = true
-systemd-networkd = true
```
Can you open an issue in twoliter to remove this feature flag if we're not using it? We should cut it from `buildsys` and `Dockerfile`.

jmt-lab replied:
yup will do

webern replied:
I opened https://github.com/bottlerocket-os/twoliter/issues/142

jmt-lab commented:
fix version number and remove accidental file

webern commented on packages/systemd/systemd.spec:
```diff
@@ -125,7 +125,6 @@ Summary: Files for networkd
 
 %package resolved
 Summary: Files for resolved
-
```
This feels unnecessary and would take the PR down to 254 files :sweat_smile:

webern commented on sources/netdog/build.rs:
```diff
@@ -0,0 +1,19 @@
+use bottlerocket_variant::{Variant, VARIANT_ENV};
```
When moving files I like to do so in an isolated commit before making changes to the files. That way we can be sure git understands that the file was moved so that changes can be tracked across the move. I was able to find at least one person on the Internet that agrees with me: [Always-move-and-rename-Git-files-in-an-isolated-commit](https://www.theserverside.com/blog/Coffee-Talk-Java-News-Stories-and-Opinions/Always-move-and-rename-Git-files-in-an-isolated-commit)

jmt-lab replied:
Will do next time, right now undoing and fixing this would be really painful

jmt-lab replied:
Fixed this

webern commented on sources/netdog/README.md:
```diff
@@ -1,6 +1,6 @@
 # netdog
 
-Current version: 0.1.0
+Current version: 0.1.1
```
If you reverted from `0.1.1` back to `0.1.0` then this needs to be regenerated/reverted as well.

webern commented on sources/netdog/src/cli/mod.rs:
```diff
@@ -4,12 +4,12 @@ pub(crate) mod node_ip;
 pub(crate) mod install;
-#[cfg(net_backend = "wicked")]
+#[cfg(feature = "wicked")]
 pub(crate) mod remove;
 
-#[cfg(net_backend = "systemd-networkd")]
+#[cfg(not(feature = "wicked"))]
```
Thank you for making this change. I think this is better instead of having mutually-exclusive features like before.

webern commented on sources/netdog/src/networkd/config/netdev.rs:
```diff
@@ -142,7 +142,7 @@ impl Display for NetDevMacAddress {
 
 impl NetDevConfig {
-    const FILE_EXT: &str = "netdev";
+    const FILE_EXT: &'static str = "netdev";
```
Unnecessary, literals/consts are always understood to be static.

```suggestion
    const FILE_EXT: &str = "netdev";
```

webern commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,92 @@
+    --target-dir=${HOME}/.cache/networkd
+%cargo_build --manifest-path %{_builddir}/sources/Cargo.toml \
+    -p netdog \
+    --features wicked \
+    --target-dir=${HOME}/.cache/wicked
```
Maybe this helps with self-documenting code: 

```suggestion
    --features wicked \
    --default-features=false \
    --target-dir=${HOME}/.cache/wicked
```

webern commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,92 @@
+posix.link("%{_cross_bindir}/netdog-wicked", "%{_cross_bindir}/netdog")
+
+%post -n %{_cross_os}netdog-systemd-networkd -p <lua>
+posix.link("%{_cross_bindir}/netdog-systemd-networkd", "%{_cross_bindir}/netdog")
```
Newline! :)
```suggestion
posix.link("%{_cross_bindir}/netdog-systemd-networkd", "%{_cross_bindir}/netdog")

```

jmt-lab commented:
Addressing comments and rebase on latest

jmt-lab committed:
netdog: move netdog folder to sources root

jmt-lab committed:
netdog: adjust netdog to use cargo feature to switch back to wicked

jmt-lab committed:
netdog: adjust packages to utilize supplement

move shared files to common

jmt-lab committed:
netdog: adjust all variant specifications

webern reviewed (dismissed):
Approved assuming that Ben's packaging suggestions are addressed.

sam-berning commented on variants/vmware-k8s-1.24/Cargo.toml:
Looks like the rebase resurrected this variant and metal-k8s-1.24 which were removed [here](https://github.com/bottlerocket-os/bottlerocket/commit/f1be358629beccaf3dc4bd18d4939ffec04a2de8)

jmt-lab replied:
yeah noticed cleaning this up

jmt-lab commented:
Addressed Ben's concerns on rpm adjustments
Rebased off develop
Restructured commits to isolate the move from edit and grouped changes by category of work

sam-berning commented on variants/vmware-dev/Cargo.toml:
```diff
@@ -33,6 +32,7 @@ included-packages = [
     "release",
     "kernel-6.1",
     "open-vm-tools",
+    "netdog-vmware-config",
```
this package doesn't exist anymore, should be removed

jmt-lab replied:
oop you are right

jmt-lab commented:
Removed redundant package in vmware variants

sam-berning reviewed (approved):
alright, lgtm pending testing. nice work!

jmt-lab commented:
Testing result on changes: Everything looks good and is functioning as expected

jmt-lab commented:
rebased off latest

bcressey commented on sources/Cargo.lock:
```diff
@@ -2661,6 +2661,7 @@ dependencies = [
 name = "netdog"
 version = "0.1.0"
 dependencies = [
+ "anyhow",
```
Please stick with `snafu` for OS components. Also, I don't think this should be part of this commit, but that's irrelevant if it goes away.

bcressey commented on sources/netdog/Cargo.toml:
```diff
@@ -8,11 +8,16 @@ publish = false
 dns-lookup = "2"
 ipnet = { version = "2", features = ["serde"] }
-imdsclient = { path = "../../imdsclient", version = "0.1" }
+imdsclient = { path = "../imdsclient", version = "0.1" }
```
There's another source of conditional compilation implied by this - `imdsclient` is used in `generate_hostname.rs` which has these:
```
#[cfg(variant_platform = "aws")]
#[cfg(not(variant_platform = "aws"))]
```

bcressey commented on sources/netdog/build.rs:
```diff
@@ -25,19 +14,6 @@ fn main() {
         }
     };
     variant.emit_cfgs();
-    emit_net_backend_cfgs();
```
`variant.emit_cfgs();` above this line is also problematic.

bcressey commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
```
Please wrap the summary line of your commit messages around 50 characters. The ellipsis in the GitHub web view is an indication that it's too long.

I tend to follow [this style guide](https://cbea.ms/git-commit/) with the addition of a "subsystem: ..." prefix to the summary.

bcressey commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+build = "../build.rs"
+
+[lib]
+path = "../packages.rs"
```
Still need to add something like:
```
[package.metadata.build-package]
source-groups = [
    "netdog"
]
```

bcressey commented on packages/os/os.spec:
```diff
@@ -511,17 +487,13 @@ install -p -m 0644 %{S:200} %{buildroot}%{_cross_tmpfilesdir}/migration.conf
 install -p -m 0644 %{S:202} %{buildroot}%{_cross_tmpfilesdir}/thar-be-updates.conf
 install -p -m 0644 %{S:203} %{buildroot}%{_cross_tmpfilesdir}/bootstrap-containers.conf
-install -p -m 0644 %{S:204} %{buildroot}%{_cross_tmpfilesdir}/netdog.conf
+
```
This extra newline still shows up in this commit.

bcressey replied:
Extra newline is still here!

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,101 @@
+License: Apache-2.0 OR MIT
+URL: https://github.com/bottlerocket-os/bottlerocket
+
+Source2: netdog-tmpfiles.conf
+
+Source10: run-netdog.mount
+Source11: write-network-status.service
+Source12: generate-network-config.service
+Source13: disable-udp-offload.service
+
+Source20: 00-resolved.conf
```
Why start with `Source2`?

Headers in a comment might help enforce the grouping, e.g.:
```
# drop-in configs
...

# units
...

# service configs
...
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,101 @@
+install -d %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+install -p -m 0644 %{S:20} %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+
+%post -n %{_cross_os}netdog-wicked -p <lua>
+posix.link("%{_cross_bindir}/netdog-wicked", "%{_cross_bindir}/netdog")
+
+%post -n %{_cross_os}netdog-systemd-networkd -p <lua>
+posix.link("%{_cross_bindir}/netdog-systemd-networkd", "%{_cross_bindir}/netdog")
```
Avoid creating absolute symlinks. They cause problems when working with different roots - e.g. when the host path is mounted in a container, or when extracting the RPM contents - since they can point outside the root to another unpackaged file.

```
%post -n %{_cross_os}netdog-wicked -p <lua>
posix.link("netdog-wicked", "%{_cross_bindir}/netdog")

%post -n %{_cross_os}netdog-systemd-networkd -p <lua>
posix.link("netdog-systemd-networkd", "%{_cross_bindir}/netdog")
```

bcressey commented on packages/os/os.spec:
```diff
@@ -478,15 +459,10 @@ install -d %{buildroot}%{_cross_unitdir}
-install -d %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
-install -p -m 0644 %{S:12} %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
-%endif
+
```
The extra newline still shows up in this commit.

bcressey commented on packages/release/Cargo.toml:
```diff
@@ -38,6 +38,7 @@ libaudit = { path = "../libaudit" }
 libgcc = { path = "../libgcc" }
 libstd-rust = { path = "../libstd-rust" }
 makedumpfile = { path = "../../packages/makedumpfile" }
+netdog = {path = "../netdog" }
```
`release` should no longer depend on `wicked` - right now this dependency is forcing it to be built for all variants, but it should only be built for variants that include it.

bcressey commented on variants/aws-ecs-1-nvidia/Cargo.toml:
```diff
@@ -20,6 +20,7 @@ included-packages = [
 # core
     "release",
     "kernel-5.10",
+    "wicked",
```
The wicked variants should also add a dependency on wicked in the `build-dependencies` section below.

jmt-lab committed:
netdog: add dogtag hostname for hostname resolution

jmt-lab committed:
netdog: move to utilize dogtag for generate-hostname and update dogtag to handle retry

jmt-lab committed:
netdog: adjust variant files and symlink

jmt-lab commented:
* Changed to utilize image-features
* Integrate use of dogtag hostname helpers

jmt-lab commented:
For testing I did the following:

- [x] aws-ecs-2 test
   - [x] verify network connectivity
   - [x] verify netdog-systemd-networkd is installed and symlinked
   - [x] verify 10-reverse-dns is installed
   - [x] verify 20-imds is installed
   - [x] verify hostname got set correctly and it did not fall through to setting ip
- [x] aws-ecs-1 test
   - [x] verify network connectivity
   - [x] verify netdog-wicked is installed and symlinked
   - [x] verify 10-reverse-dns is installed
   - [x] verify 20-imds is installed
   - [x] verify hostname got set correctly and it did not fallthrough to setting ip

bcressey commented on variants/aws-k8s-1.24/Cargo.toml:
```diff
@@ -39,3 +39,4 @@ cni-plugins = { path = "../../packages/cni-plugins" }
 kernel-5_15 = { path = "../../packages/kernel-5.15" }
 kubernetes-1_24= { path = "../../packages/kubernetes-1.24" }
 release = { path = "../../packages/release" }
+wicked = { path = "../../packages/wicked" }
```
We shouldn't need this change.

bcressey commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+[build-dependencies]
+glibc = { path = "../glibc" }
+
+[dependencies]
+wicked = { path = "../wicked" }
```
Just because it's weird to only depend on one of the two backend daemons:
```suggestion
[dependencies]
wicked = { path = "../wicked" }
systemd = { path = "../systemd" }
```

bcressey commented on sources/dogtag/src/lib.rs:
```diff
@@ -28,7 +29,7 @@ pub struct Cli {
 
 /// find_hostname will utilize the helpers located in /var/bottlerocket/dogtag/ to try and discover the hostname
-pub async fn find_hostname() -> Result<String> {
+pub async fn find_hostname(ip_addr: IpAddr) -> Result<String> {
```
nit: comment still refers to `/var/bottlerocket/dogtag`.

bcressey commented on sources/netdog/src/cli/generate_hostname.rs:
```diff
@@ -0,0 +1,30 @@
+        .context(error::CurrentIpReadFailedSnafu { path: CURRENT_IP })?;
+    let ip = IpAddr::from_str(&ip_string).context(error::IpFromStringSnafu { ip: &ip_string })?;
+
+    // If no hostname has been determined we return the IP address of the host, replacing invalid ipv6 chars.
+    let hostname = find_hostname(ip)
+        .await
+        .unwrap_or(ip_string.replace(':', "-"));
```
Might be better to handle this in `dogtag`.

The other potential edge case is where we "find" a hostname, but it's the same as the IPv6 address we passed in, in which case we'd still want to do the replacement.

jmt-lab replied:
So the dogtag lib accepts a full IpAddr object instead of the ip_string itself. I imagined this replacement was more needed for storage of the hostname in the settings or on disk configuration?

jmt-lab replied:
We talked about it going to move the ip-address fallback in dogtag find_hostname

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+
+%package -n %{_cross_os}netdog-systemd-networkd
+Summary: Bottlerocket network configuration helper
+Provides: %{_cross_os}netdog = 2
```
We don't need this.

```suggestion
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+
+%package -n %{_cross_os}netdog-wicked
+Summary: Bottlerocket network configuration helper
+Provides: %{_cross_os}netdog = 1
```
```suggestion
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%description
+%{summary}.
+
+%package -n %{_cross_os}netdog-common
```
Prefer the shorthand syntax for subpackages that suffix the main package name:
```suggestion
%package common
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Source20: 00-resolved.conf
+
+BuildRequires: %{_cross_os}glibc-devel
+Requires: %{_cross_os}netdog
```
This is a little weird - package requiring itself - and isn't needed when we drop the epoch provides hack:
```suggestion
```

The two requirements below are a nicer way of doing that.

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+
+BuildRequires: %{_cross_os}glibc-devel
+Requires: %{_cross_os}netdog
+Requires: (%{_cross_os}netdog-systemd-networkd if %{_cross_os}image-feature(systemd-networkd))
+Requires: (%{_cross_os}netdog-wicked if %{_cross_os}image-feature(no-systemd-networkd))
```
shorthand syntax:
```suggestion
Requires: (%{name}-systemd-networkd if %{_cross_os}image-feature(systemd-networkd))
Requires: (%{name}-wicked if %{_cross_os}image-feature(no-systemd-networkd))
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Summary: Common configuration for Bottlerocket's network configuration helper
+Requires: %{_cross_os}hostname-reverse-dns
+Requires: (%{_cross_os}hostname-imds if %{_cross_os}variant-platform(aws))
+%description -n %{_cross_os}netdog-common
```
```suggestion
%description common
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%description -n %{_cross_os}netdog-common
+%{summary}.
+
+%package -n %{_cross_os}netdog-systemd-networkd
```
```suggestion
%package systemd-networkd
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%package -n %{_cross_os}netdog-systemd-networkd
+Summary: Bottlerocket network configuration helper
+Provides: %{_cross_os}netdog = 2
+Requires: %{_cross_os}netdog-common
```
```suggestion
Requires: %{name}-common
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Requires: %{_cross_os}systemd-networkd
+Requires: %{_cross_os}systemd-resolved
+Requires: %{_cross_os}netdog
+Conflicts: (%{_cross_os}netdog-wicked or %{_cross_os}image-feature(no-systemd-networkd))
+%description -n %{_cross_os}netdog-systemd-networkd
```
```suggestion
Conflicts: (%{name}-wicked or %{_cross_os}image-feature(no-systemd-networkd))
%description systemd-networkd
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%description -n %{_cross_os}netdog-systemd-networkd
+%{summary}.
+
+%package -n %{_cross_os}netdog-wicked
```
```suggestion
%package wicked
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%package -n %{_cross_os}netdog-wicked
+Summary: Bottlerocket network configuration helper
+Provides: %{_cross_os}netdog = 1
+Requires: %{_cross_os}netdog-common
```
```suggestion
Requires: %{name}-common
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Requires: %{_cross_os}netdog-common
+Requires: %{_cross_os}systemd-networkd
+Requires: %{_cross_os}systemd-resolved
+Requires: %{_cross_os}netdog
```
```suggestion
Requires: %{name}
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Provides: %{_cross_os}netdog = 1
+Requires: %{_cross_os}netdog-common
+Requires: %{_cross_os}wicked
+Requires: %{_cross_os}netdog
```
```suggestion
Requires: %{name}
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Requires: %{_cross_os}netdog-common
+Requires: %{_cross_os}wicked
+Requires: %{_cross_os}netdog
+Conflicts: (%{_cross_os}netdog-systemd-networkd or %{_cross_os}image-feature(systemd-networkd))
```
```suggestion
Conflicts: (%{name}-systemd-networkd or %{_cross_os}image-feature(systemd-networkd))
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+Requires: %{_cross_os}wicked
+Requires: %{_cross_os}netdog
+Conflicts: (%{_cross_os}netdog-systemd-networkd or %{_cross_os}image-feature(systemd-networkd))
+%description -n %{_cross_os}netdog-wicked
+%{summary}.
```
```suggestion
%description wicked
%{summary}.
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+    --target-dir=${HOME}/.cache/wicked
+
+%install
+install -d %{buildroot}%{_cross_libexecdir}/hostname-resolvers
```
I would pick a consistent phrase - either "hostname resolvers" or "hostname detectors" ("detectors" is in the package description and would be easiest to change).

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+install -d %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+install -p -m 0644 %{S:20} %{buildroot}%{_cross_libdir}/systemd/resolved.conf.d
+
+%post -n %{_cross_os}netdog-wicked -p <lua>
+posix.symlink("netdog-wicked", "%{_cross_bindir}/netdog")
+
+%post -n %{_cross_os}netdog-systemd-networkd -p <lua>
+posix.symlink("netdog-systemd-networkd", "%{_cross_bindir}/netdog")
```
```suggestion
%post wicked -p <lua>
posix.symlink("netdog-wicked", "%{_cross_bindir}/netdog")

%post systemd-networkd -p <lua>
posix.symlink("netdog-systemd-networkd", "%{_cross_bindir}/netdog")
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%files -n %{_cross_os}hostname-imds
+%{_cross_libexecdir}/hostname-resolvers/20-imds
+
+%files -n %{_cross_os}netdog-common
```
```suggestion
%files common
```

bcressey replied:
Actually, you don't have anything in the `netdog` package now, and we don't need to play the weird game with epochs, so I would just drop the "common" subpackage and move this stuff into `%files` (for the main netdog package).

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,133 @@
+%{_cross_unitdir}/disable-udp-offload.service
+%{_cross_unitdir}/run-netdog.mount
+
+%files -n %{_cross_os}netdog-systemd-networkd
+%{_cross_bindir}/netdog-systemd-networkd
+%{_cross_unitdir}/write-network-status.service
+%dir %{_cross_libdir}/systemd/resolved.conf.d
+%{_cross_libdir}/systemd/resolved.conf.d/00-resolved.conf
+
+%files -n %{_cross_os}netdog-wicked
+%{_cross_bindir}/netdog-wicked
```
```suggestion
%files systemd-networkd
%{_cross_bindir}/netdog-systemd-networkd
%{_cross_unitdir}/write-network-status.service
%dir %{_cross_libdir}/systemd/resolved.conf.d
%{_cross_libdir}/systemd/resolved.conf.d/00-resolved.conf

%files wicked
%{_cross_bindir}/netdog-wicked
```

jmt-lab commented:
Updated dogtag to handle fallback to ip address

sam-berning commented on variants/metal-k8s-1.27/Cargo.toml:
```diff
@@ -16,6 +16,7 @@ partition-plan = "unified"
 [package.metadata.build-variant.image-features]
 grub-set-private-var = true
 unified-cgroup-hierarchy = true
+systemd-networkd = true
```
why is this present in only this one variant?

jmt-lab replied:
good catch i've removed it

jmt-lab commented:
Fix formatting

bcressey commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+
+[dependencies]
+wicked = { path = "../wicked" }
+systemd = { path = "../systemd" }
```
Should be a newline at the end of this file - can you verify that your editor is configured to do this?

bcressey commented on packages/netdog/Cargo.toml:
```diff
@@ -0,0 +1,23 @@
+[package.metadata.build-packages]
+source-groups = [
+    "netdog",
+    "dogtag",
```
nit: we could put `dogtag` under `netdog` (like how various update-related crates are under `sources/updater`) but that can be done later

bcressey commented on sources/netdog/src/cli/generate_hostname.rs:
```diff
@@ -0,0 +1,30 @@
+/// Generate hostname from installed hostname resolvers
+pub(crate) struct GenerateHostnameArgs {}
+
+/// Attempt to resolve assigned IP address, if unsuccessful use the IP as the hostname.
+///
+/// The result is returned as JSON. (intended for use as a settings generator)
+pub(crate) async fn run() -> Result<()> {
+    let ip_string = fs::read_to_string(CURRENT_IP)
+        .context(error::CurrentIpReadFailedSnafu { path: CURRENT_IP })?;
+    let ip = IpAddr::from_str(&ip_string).context(error::IpFromStringSnafu { ip: &ip_string })?;
+
+    // If no hostname has been determined we return the IP address of the host, replacing invalid ipv6 chars.
+    let hostname = find_hostname(ip)
+        .await
+        .context(error::HostnameDetectionSnafu)?;
```
nit: comments are out-of-date since they refer to a `dogtag` implementation detail:
```suggestion
/// Attempt to convert the assigned IP address to a valid hostname.
///
/// The result is returned as JSON. (intended for use as a settings generator)
pub(crate) async fn run() -> Result<()> {
    let ip_string = fs::read_to_string(CURRENT_IP)
        .context(error::CurrentIpReadFailedSnafu { path: CURRENT_IP })?;
    let ip = IpAddr::from_str(&ip_string).context(error::IpFromStringSnafu { ip: &ip_string })?;
    let hostname = find_hostname(ip)
        .await
        .context(error::HostnameDetectionSnafu)?;
```

bcressey commented on packages/netdog/netdog.spec:
```diff
@@ -0,0 +1,123 @@
+
+%package systemd-networkd
+Summary: Bottlerocket network configuration helper
+Requires: %{_cross_os}netdog
```
```suggestion
Requires: %{name}
```

jmt-lab commented:
* Rebased off develop
* Bump netdog version
* Adjust Requires
* Add missing newline
* Remove extra newline
* Adjust comment

bcressey reviewed (approved):
🚢 🚢 🚢 🥅 🐶
